forge/src/multi_runner.rs

```diff
@@ -191,13 +191,51 @@ pub struct MultiContractRunner {
 }
 
 impl MultiContractRunner {
-    pub fn test(
+    pub fn test_stream(
         &mut self,
         filter: &(impl TestFilter + Send + Sync),
+        stream_result: impl Fn(String, BTreeMap<String, TestResult>) + Sync,
     ) -> Result<BTreeMap<String, BTreeMap<String, TestResult>>> {
-        // TODO: Convert to iterator, ideally parallel one?
         let contracts = std::mem::take(&mut self.contracts);
+        let vicinity = self.evm_opts.vicinity()?;
+        let backend = self.evm_opts.backend(&vicinity)?;
+        let source_paths = self.source_paths.clone();
 
+        let results = contracts
+            .par_iter()
+            .filter(|(name, _)| filter.matches_path(source_paths.get(*name).unwrap()))
+            .filter(|(name, _)| filter.matches_contract(name))
+            .map(|(name, (abi, deploy_code, libs))| {
+                // unavoidable duplication here?
+                let result = match backend {
+                    BackendKind::Simple(ref backend) => {
+                        self.run_tests(name, abi, backend, deploy_code.clone(), libs, filter)?
+                    }
+                    BackendKind::Shared(ref backend) => {
+                        self.run_tests(name, abi, backend, deploy_code.clone(), libs, filter)?
+                    }
+                };
+                Ok((name.clone(), result))
+            })
+            .filter_map(|x: Result<_>| x.ok())
+            .filter_map(
+                |(name, result)| if result.is_empty() { None } else { Some((name, result)) },
+            )
```

Comment:
If you add an `.enumerate()` here you get an `index` and you can remove the `Mutex` in cli/src/cmd/test.rs

Reply:
Mutex needs to be there in some form so tests are grouped by contract

Reply:
They already are, no? `run_tests` creates a `ContractRunner` and returns a list of test results for the entire contract, which you then pass back to the streaming function

Reply:
Right, but when does it return? `par_iter` runs on multiple threads. Entirely possible for output of one more contracts to get interleaved.

Reply:
Ah, right, in terms of outputting to stdout. Makes sense!